src/overlays/actors/ovl_En_Grasshopper/z_en_grasshopper.c

```diff
@@ -15,41 +16,114 @@ void EnGrasshopper_Destroy(Actor* thisx, PlayState* play);
-    /* Powder Keg     */ DMG_ENTRY(1, 0xE),
+void EnGrasshopper_ChangeAnim(EnGrasshopper* this, s32 animIndex);
+void EnGrasshopper_RaiseTail(EnGrasshopper* this);
+void EnGrasshopper_LowerTail(EnGrasshopper* this);
+void EnGrasshopper_DecideAction(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupFly(EnGrasshopper* this);
+void EnGrasshopper_Fly(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_RoamInCircles(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupBank(EnGrasshopper* this);
+void EnGrasshopper_Bank(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupBounced(EnGrasshopper* this);
+void EnGrasshopper_Bounced(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupApproachPlayer(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_ApproachPlayer(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupAttack(EnGrasshopper* this);
+void EnGrasshopper_Attack(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupWaitAfterAttack(EnGrasshopper* this);
+void EnGrasshopper_WaitAfterAttack(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_Damaged(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_Dead(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_SetupFall(EnGrasshopper* this);
+void EnGrasshopper_Fall(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_InitializeEffect(EnGrasshopper* this, Vec3f* pos);
+void EnGrasshopper_UpdateEffects(EnGrasshopper* this, PlayState* play);
+void EnGrasshopper_DrawEffects(EnGrasshopper* this, PlayState* play);
+
+typedef enum {
+    /*  0 */ EN_GRASSHOPPER_ACTION_FLY,
+    /*  1 */ EN_GRASSHOPPER_ACTION_ROAM_IN_CIRCLES,
+    /*  2 */ EN_GRASSHOPPER_ACTION_BANK,
+    /*  3 */ EN_GRASSHOPPER_ACTION_BOUNCED,
+    /*  4 */ EN_GRASSHOPPER_ACTION_DECIDE_ACTION,
+    /*  5 */ EN_GRASSHOPPER_ACTION_APPROACH_PLAYER,
+    /*  6 */ EN_GRASSHOPPER_ACTION_ATTACK,
+    /*  7 */ EN_GRASSHOPPER_ACTION_WAIT_AFTER_ATTACK,
+    /*  8 */ EN_GRASSHOPPER_ACTION_DAMAGED,
+    /*  9 */ EN_GRASSHOPPER_ACTION_DEAD,
+    /* 10 */ EN_GRASSHOPPER_ACTION_FALL,
+} EnGrasshopperAction;
+
+typedef enum {
+    /* 0 */ EN_GRASSHOPPER_DECISION_ATTACK,
+    /* 1 */ EN_GRASSHOPPER_DECISION_FLY,
+    /* 2 */ EN_GRASSHOPPER_DECISION_ROAM_IN_CIRCLES, // Never used in the final game
+} EnGrasshopperNextAction;
+
+typedef enum {
+    /* 0 */ EN_GRASSHOPPER_BANK_STATE_BANKING,
+    /* 1 */ EN_GRASSHOPPER_BANK_STATE_DONE,
+} EnGrasshopperBankState;
+
+static s32 sOccupiedIndices[] = {
+    false, false, false, false, false,
+};
+
+static s8 sLimbIndexToShadowBodyPartsIndex[] = {
+    -1, -1, 0, 1, 2, 3, 4, 5, -1, -1, 6, 7, -1, 8, 9, -1, 10, 11, -1, 12, 13, -1, -1, -1,
+};
+
+static s8 sParentBodyParts[] = {
+    -1, -1, -1, 0, 1, 2, 3, -1, -1, -1, -1, 6, -1, -1, 8, -1, -1, 10, -1, -1, 12, -1, -1, -1,
+};
+
+static u8 sShadowSizes[] = {
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+};
```

Comment:
Do all of these arrays have to be the same size, dependent on something? If so might be best to initialise with a specified size.

Reply:
`sLimbIndexToShadowBodyPartsIndex` has to be `DRAGONFLY_LIMB_MAX`-sized. The other two I'm not sure about.

Reply:
Oh, never mind, I remember the other two now. This is a weird one because they get used like this:
```c
SubS_GenShadowTex(this->shadowBodyPartsPos, &this->actor.world.pos, shadowTex, i / 5.0f,
    ARRAY_COUNT(this->shadowBodyPartsPos), sShadowSizes, sParentBodyParts);
```
The thing is that `this->shadowBodyPartsPos` is defined like this:
```c
/* 0x03D8 */ Vec3f shadowBodyPartsPos[14]; // Was almost certainly declared with 24 entries, but only the first 14 are used.
/* 0x0480 */ UNK_TYPE1 unk_480[0x78];
```
So the other two arrays are length 24 now, which theoretically *should* be the size of `this->shadowBodyPartsPos`. But for some bizarre reason, they only use the first 14 entries.

Reply:
I'd be tempted to beef that up to the theoretical extent and use that. Up to you, this is such a messy actor it's hard to decide.

Reply:
Yeah, I asked Anon about this, as the subs expert, and he suggested the current approach. But it could just be:
```c
SubS_GenShadowTex(this->shadowBodyPartsPos, &this->actor.world.pos, shadowTex, i / 5.0f,
    14, sShadowSizes, sParentBodyParts);
```
With `shadowBodyPartsPos` being length 24.

Reply:
When we discussed in Discord, I don't think I realized those other arrays were also length 24. So yeah, I think it would be better to have the shadow body parts array the full 24, than just hardcode the 14, probably with a comment.

Reply:
If you want to see the original conversation with Anon, it starts here: https://discord.com/channels/688807550715560050/688851976746041391/1009998644637872238

Reply:
Ah, I think we commented at the same time. I'll switch to the length 24 approach.